DOCSP-31621: socks dependency is optional (#748)
mongodb/docs-node: source/fundamentals/connection/socks.txt

.. _node-connect-socks:

===========================
Enable SOCKS5 Proxy Support
===========================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

In this guide, you can learn how to connect to MongoDB instances by using
a SOCKS5 proxy. SOCKS5 is a standardized protocol for connecting
to network services through a proxy server.

.. tip::
   
   To learn more about the SOCKS5 protocol, see the Wikipedia entry on
   :wikipedia:`SOCKS <w/index.php?title=SOCKS&oldid=1160087146>`.

Install the socks Package
-------------------------

> Starting in version 6.0 of the {+driver-short+}, you must *manually* install

Reviewer: Suggestion:
I think "manually" is unnecessary since there isn't a contrasting option that the user can perform. E.g. "You must automatically install" is not a possible action.

.. _node-connect-socks:

===========================
Enable SOCKS5 Proxy Support
===========================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

In this guide, you can learn how to connect to MongoDB instances with
SOCKS5 proxy support. SOCKS5 is a standardized protocol for connecting
to network services through a separate proxy server.

.. tip::
   
   To learn more about SOCKS, see the Wikipedia entry on

>    :wikipedia:`SOCKS <SOCKS>`.

Author: This was the URL I got when I clicked Permanent URL in the tools tab:

> https://en.wikipedia.org/w/index.php?title=SOCKS&oldid=1160087146

is this what I should use?

>    :wikipedia:`SOCKS <SOCKS>`.

Reviewer: Yes, please use that. To get that URL from the wikipedia page, you can click "Tools > Permanent Link".

> Load the socks Module

Reviewer: Suggestion:

I think using consistent and precise terminology for this would help the reader understand the task.

I think this should be "Install the socks Package" or "Install the socks Module". I think packages and modules are technically very similar, but package is more specific in that it's "packaged" using a package.json file.

See https://docs.npmjs.com/about-packages-and-modules#about-modules for more info.

---------------------

Starting in version 6.0 of the {+driver-short+}, you must *manually* load
the ``socks`` dependency to use SOCKS5 proxy support in your
application. You can install ``socks`` by running the following command
in your terminal:

.. code-block:: bash
   
   npm i socks

SOCKS5 Client Options
---------------------

You can set options in your ``MongoClientOptions`` instance or
parameters in your connection URI to configure SOCKS5 proxy support for 
your connection. The following table describes the client options
related to SOCKS5.

.. list-table::
   :header-rows: 1
   :widths: 1 1 1 1

   * - Name
     - Accepted Values
     - Default Value
     - Description

   * - **proxyHost**
     - string
     - ``null``

>      - Specifies the IPv4/IPv6 address or domain name on which a SOCKS5
>        proxy is listening.

Reviewer: Issue:

From [Slashes](https://www.mongodb.com/docs/meta/style-guide/style/punctuation/) in the style guide:
> Don't use a slash mark (/) to present a choice among, or a series of, actions or objects. Rewrite the phrase to eliminate the slash mark. Exceptions are established terms like client/server and read/write.

I think it could also be helpful to remove the perspective of the "proxy" which is "listening" since the configuration specifies behavior for the driver.

Suggestion:

```suggestion
     - Specifies the SOCKS5 proxy IPv4 address, IPv6 address, or domain name.
```

   * - **proxyPort**
     - non-negative integer
     - ``null``
     - Specifies the TCP port number of the SOCKS5 proxy server specified in the
       ``proxyHost`` option. If you have set the ``proxyHost`` option,
       this value of this option defaults to ``1080``.

   * - **proxyUsername**
     - string
     - ``null``

>      - Specifies the username for username and password authentication to the SOCKS5

Reviewer: Suggestion:

I think the sentence is a little confusing since it presents "username" and "username and password". I understand this is a type of authentication, but I don't think any other type that might cause confusion has been mentioned on this page.

```suggestion
     - Specifies the username for authentication to the SOCKS5
```

>        proxy server specified in the ``proxyHost`` option. If you set

Reviewer: Suggestion:

I think it's important to identify it's for the proxy server, but think it's reasonable to expect the user to understand that the proxy is the one specified in the proxyHost option.

```suggestion
       proxy server. If you set
```

>        this option to a zero-length string, the driver ignores it.

Reviewer: Question:

I wasn't able to determine what "ignores it" entailed. I am guessing that if either or both the proxyUsername and proxyPassword values are null (default) or zero-length strings, that the driver does not send any authentication information? Or that if I only passed a username, that it would present that for authentication, but no password and vice versa?

Author: Yes, if the option is set to an empty string, the driver does not acknowledge that the option is set. I'm not sure what happens if you set one or the other -- I am assuming you'd receive a runtime error

Reviewer: Ah I see, there's no validation or acknowledgement. The value just gets passed as a parameter, but if the value is "", that gets changed to undefined since it evaluates to false in these statements:

https://github.com/mongodb/node-mongodb-native/blob/v5.8.0/src/cmap/connect.ts#L468-L469

No action needed.

   * - **proxyPassword**
     - string
     - ``null``
     - Specifies the password for username and password authentication to the SOCKS5
       proxy server specified in the ``proxyHost`` option. If you set
       this option to a zero-length string, the driver ignores it.

.. important::
   

>    The driver will throw an error if you set the ``proxyPort``,

Reviewer: Issue:
[Use Present Tense](https://www.mongodb.com/docs/meta/style-guide/writing/use-present-tense/).

Suggestion:
```suggestion
   The driver throws an error if you set the ``proxyPort``,
```

     proxyPort: 1010,
     proxyUsername: "<username>",
     proxyPassword: "<password>",
   };
   
   // Create a client with the socksOptions object
   const client = new MongoClient(uri, socksOptions);
   
   async function run() {
     try {
       const db = client.db("myDB");
       const myColl = db.collection("myColl");
       const doc = await myColl.findOne({});
       console.log(doc);
     } finally {
       await client.close();
     }
   }
   run().catch(console.dir);

Additional Information
----------------------

For more information about SOCKS5 proxy support, see the
`MongoDB SOCKS5 specification <https://github.com/mongodb/specifications/blob/master/source/socks5-support/socks5.rst>`__.

API Documentation
~~~~~~~~~~~~~~~~~

- `MongoClientOptions <{+api+}/interfaces/MongoClientOptions.html>`__
- `MongoClient <{+api+}/classes/MongoClient.html>`__
- `ProxyOptions <{+api+}/interfaces/ProxyOptions.html>`__
- `proxyHost client option <{+api+}/interfaces/ProxyOptions.html#proxyHost>`__
- `proxyPassword client option <{+api+}/interfaces/ProxyOptions.html#proxyPassword>`__
- `proxyPort client option <{+api+}/interfaces/ProxyOptions.html#proxyPort>`__

> - `proxyUsername client option <{+api+}/interfaces/ProxyOptions.html#proxyUsername>`__

Reviewer: Suggestion:
I think these can be removed since they are on the ProxyOptions page and all the information is densely located.
